base/mmap.jl

```diff
@@ -63,106 +151,70 @@ function mmap_stream_settings(s::IO)
-        ro = isreadonly(s)
-        create = true
-    else
-        # shared memory
-        hdl = -1
-        name = utf16(s.name)
-        ro = s.readonly
-        create = s.create
+        n *= d
     end
     len = prod(dims)*sizeof(T)
-    const granularity::Int = ccall(:jl_getallocationgranularity, Clong, ())
-    if len < 0
-        throw(ArgumentError("requested size must be ≥ 0, got $len"))
-    end
-    if len > typemax(Int)-granularity
-        throw(ArgumentError("file is too large ot memory-map on this platform"))
-    end
-    # Set the offset to a page boundary
-    offset_page::FileOffset = div(offset, granularity)*granularity
-    szfile = convert(Csize_t, len + offset)
-    szarray = szfile - convert(Csize_t, offset_page)
-    access = ro ? 4 : 2
-    if create
-        flprotect = ro ? 0x02 : 0x04
-        mmaphandle = ccall(:CreateFileMappingW, stdcall, Ptr{Void}, (Cptrdiff_t, Ptr{Void}, Cint, Cint, Cint, Cwstring),
-            hdl, C_NULL, flprotect, szfile>>32, szfile&typemax(UInt32), name)
-    else
-        mmaphandle = ccall(:OpenFileMappingW, stdcall, Ptr{Void}, (Cint, Cint, Cwstring),
-            access, true, name)
-    end
-    if mmaphandle == C_NULL
-        error("could not create file mapping: $(FormatMessage())")
-    end
-    viewhandle = ccall(:MapViewOfFile, stdcall, Ptr{Void}, (Ptr{Void}, Cint, Cint, Cint, Csize_t),
-        mmaphandle, access, offset_page>>32, offset_page&typemax(UInt32), szarray)
-    if viewhandle == C_NULL
-        error("could not create mapping view: $(FormatMessage())")
-    end
-    A = pointer_to_array(convert(Ptr{T}, viewhandle+offset-offset_page), dims)
-    finalizer(A, x->Libc.munmap(viewhandle, mmaphandle))
-    return A
-end
+    @unix_only begin
+        isa(s,SharedMemSpec) && throw(ArgumentError("$s is not valid on unix"))
+        prot, flags, iswrite = settings(s)
+        if iswrite && grow
+            grow!(len, fd(s), offset)
+        end
+        mm = Stream(len, prot, flags, fd(s), offset, isdefined(s,:name) ? s.name : "")
+    end # @unix_only
 
+    @windows_only begin
+        if isa(s,IO)
+            hdl::Int = Base._get_osfhandle(RawFD(fd(s))).handle
+            hdl == -1 && throw(ArgumentError("could not get handle for file to map: $(Base.FormatMessage())"))
+            name = Ptr{Cwchar_t}(C_NULL)
+            ro = isreadonly(s)
+            create = true
+        else
+            # shared memory
+            hdl = -1
+            name = utf16(s.name)
+            ro = s.readonly
+            create = s.create
+        end
+        mm = Stream(len, hdl, ro, create, offset, name)
+    end # @windows_only
+
+    A = pointer_to_array(convert(Ptr{T}, UInt(mm.viewhandle)+mm.offset), dims)
+    return A
 end
 
 # Mmapped-bitarray constructor
-function mmap_bitarray{N}(dims::NTuple{N,Integer}, s::IOStream, offset::FileOffset)
+mmap_array{N}(::Type{Bool}, dims::NTuple{N,Integer}, s::IOStream, offset::FileOffset) =
+    mmap_bitarray(dims, s, offset)
```

Comment:
This will cause trouble for anyone who is actually wanting to mmap an array of bools, i.e., UInt8 with values of 0 or 1.